crates/wasi-provider/src/wasi_runtime.rs

```diff
@@ -71,64 +88,246 @@ impl WasiRuntime {
+                                finished_at: Some(Time(chrono::Utc::now())),
+                                ..Default::default()
+                            }),
+                            ..Default::default()
+                        })
+                        .expect("status should be able to send");
+                    // Converting from anyhow
+                    Err(format_err!("{}: {}", message, e))
                 }
-            })
-            .collect::<Result<Vec<_>, _>>()?;
-
-        let instance = wasmtime::Instance::new(&module, &imports)
-            .map_err(|e| format_err!("unable to instantiate module: {}", e))?;
-
-        // NOTE(taylor): In the future, if we want to pass args directly, we'll
-        // need to do a bit more to pass them in here.
-        info!("starting run of module");
-        instance
-            .get_export("_start")
-            .expect("export")
-            .func()
-            .unwrap()
-            .call(&[])
-            .unwrap();
-
-        info!("module run complete");
-        Ok(())
+            }?;
+
+            info!("module run complete");
+            status_sender
+                .broadcast(ContainerState {
+                    terminated: Some(ContainerStateTerminated {
+                        message: Some("Module run completed".into()),
+                        exit_code: 0,
+                        finished_at: Some(Time(chrono::Utc::now())),
+                        ..Default::default()
+                    }),
+                    ..Default::default()
+                })
+                .expect("status should be able to send");
+            Ok(())
+        });
+
+        Ok(RuntimeHandle::new(
+            tokio::fs::File::from_std(self.output.reopen()?),
+            handle,
+            status_recv,
+        ))
+    }
+}
+
+/// Represents a handle to a running WASI instance. Right now, this is
+/// experimental and just for use with the [crate::WasiProvider]. If we like
+/// this pattern, we will expose it as part of the kubelet crate
+pub struct RuntimeHandle<R: AsyncReadExt + AsyncSeekExt + Unpin> {
+    output: BufReader<R>,
+    handle: JoinHandle<Result<(), failure::Error>>,
+    status_channel: Receiver<ContainerState>,
+}
+
+impl<R: AsyncReadExt + AsyncSeekExt + Unpin> RuntimeHandle<R> {
+    /// Create a new handle with the given reader for log output and a handle to
+    /// the running tokio task. The sender part of the channel should be given
+    /// to the running process and the receiver half passed to this constructor
+    /// to be used for reporting current status
+    pub fn new(
+        output: R,
+        handle: JoinHandle<Result<(), failure::Error>>,
+        status_channel: Receiver<ContainerState>,
+    ) -> Self {
+        RuntimeHandle {
+            output: BufReader::new(output),
+            handle,
+            status_channel,
+        }
+    }
+
+    pub async fn output(&mut self) -> Result<Vec<u8>, failure::Error> {
```

Comment:
Thinking out loud. I wonder if it will make sense for this method to take the buffer it will write into instead of creating a new one.

Reply:
Oooh yes, much better idea

Reply:
Yep, this was a nicer way of doing it